crates/biome_js_analyze/src/semantic_analyzers/correctness/use_hook_at_top_level.rs

```diff
@@ -85,40 +83,108 @@ declare_rule! {
     }
 }
 
+#[derive(Clone)]
+enum AnyJsFunctionOrMethod {
+    Function(AnyJsFunction),
+    ObjectMethod(JsMethodObjectMember),
```

Comment:
It could be argued I should add a variant for `ClassMethod` too here, but the reason I didn't is because implementing components inside an object literal still seems like a somewhat reasonable use case if you want to use the object as a lookup table for which component to use (and indeed it was reported by a user). But implementing them inside class methods seems a horrible idea in my opinion. Anyone doing that would need to create class instances before being able to invoke the component, which means you may run into issues with component identities and need to be very careful to memoize the class instances at least. Then there's the issue class methods would expect `this` to be set correctly, but React won't do this for you when invoking the component. All of that can be worked around and it's technically possible, but it seems far-fetched enough to me that I would wait to hear if any user even expects this to work.